ocaml/lambda/matching.ml

```diff
@@ -129,65 +129,30 @@ let may_compat = MayCompat.compat
 
 and may_compats = MayCompat.compats
 
-(* The free variables in a guarded rhs can be precomputed to be used in an
-   optimization, or uncomputed, in which case the optimization is not applied.
-*)
-type guarded_free_variables =
-  | Precomputed of Ident.Set.t
-  | Uncomputed
-
-let map_guarded_free_variables ~f = function
-  | Precomputed free_variables -> Precomputed (f free_variables)
-  | Uncomputed -> Uncomputed
-
 type rhs =
-  | Guarded of
-      { patch_guarded: patch:lambda -> lambda
-      ; free_variables: guarded_free_variables }
-  (* Guarded rhs's must allow for fallthrough if the guard fails.
-
-     When translating a guarded rhs, the code to execute on fallthrough must
-     be "patched" in. Because the fallthrough code is translated after the
-     rhs is created, guarded actions rhs's a function [patch_guarded], which
-     generates a lambda term for the action from the fallthrough code.
-
-     One translation optimization requires us to check the free variables of an
-     rhs. When it is simple to do so, we precompute these free variables for use
-     in the optimization. Otherwise, we leave them uncomputed and do not apply
-     the optimization.
+  | Guarded of { action: lambda; exn_label: static_label }
+  (* Guarded rhs's must allow for fallthrough if the guard fails. This
+     fallthrough is facilitated by raising with label [exn_label].
   *)
   | Unguarded of lambda
 
-let mk_boolean_guarded_rhs ~patch_guarded ~free_variables =
-  Guarded { patch_guarded; free_variables = Precomputed free_variables }
+let lambda_of_rhs = function
+  | Unguarded action | Guarded { action; _ } -> action
```

Comment:
I'd actually prefer to get rid of this function entirely and push any pattern-matching out to clients. This client is the one who has enough context to decide whether it's appropriate to treat the `action` as the "whole lambda term".

E.g. what if a client wanted to collect "unhandled static raise labels"? They'd get the wrong answer if they used `lambda_of_rhs`.

It's obvious that collecting free variables is ok to do, but that is clearest to see by inlining the match at the point we collect the free variables.